pkg/pillar/cmd/tpmmgr/tpmmgr.go

```diff
@@ -870,6 +873,100 @@ func testEcdhAES() error {
 	return nil
 }
 
+// DecryptCipherBlock : Decryption API, for encrypted object information received from controller
+func DecryptCipherBlock(cipherBlock types.CipherBlock) ([]byte, error) {
+	// TBD:XXX, for nodes not having tpm chip, the device private key can be used
+	// which can be wrqpped up inside DecryptWithEcdhKey
+	if !IsTpmEnabled() {
+		return []byte{}, errors.New("Not supported")
+	}
+	if len(cipherBlock.CipherData) == 0 {
+		return []byte{}, errors.New("Invalid Cipher Payload")
+	}
+	switch cipherBlock.KeyExchangeScheme {
+	case zconfig.KeyExchangeScheme_KEA_NONE:
+		return []byte{}, errors.New("No Key Exchange Scheme")
+
+	case zconfig.KeyExchangeScheme_KEA_ECDH:
+		clearData, err := decryptCipherBlockWithECDH(cipherBlock)
+		if err == nil {
+			if ret := validateDataHash(clearData,
+				cipherBlock.ClearTextHash); !ret {
+				return []byte{}, errors.New("Data Validation Failed")
+			}
+			return clearData, nil
+		}
+	}
+	return []byte{}, errors.New("Unsupported Cipher Key Exchange Scheme")
+}
+
+func decryptCipherBlockWithECDH(cipherBlock types.CipherBlock) ([]byte, error) {
+	if len(cipherBlock.ControllerCert) == 0 {
+		return []byte{}, errors.New("No Peer Public Certficate")
+	}
+	cert, err := getControllerCertEcdhKey(cipherBlock)
+	if err != nil {
+		log.Errorf("Could not extract ECDH Certificate Information")
+		return []byte{}, err
+	}
+
+	switch cipherBlock.EncryptionScheme {
+	case zconfig.EncryptionScheme_SA_NONE:
+		return []byte{}, errors.New("No Encryption")
+
+	case zconfig.EncryptionScheme_SA_AES_256_CFB:
+		if len(cipherBlock.InitialValue) == 0 {
+			return []byte{}, errors.New("Invalid Initial value")
+		}
+		clearData := make([]byte, len(cipherBlock.CipherData))
+		err = DecryptSecretWithEcdhKey(cert.X, cert.Y,
+			cipherBlock.InitialValue, cipherBlock.CipherData, clearData)
+		if err != nil {
+			log.Errorf("Decryption failed with error %v\n", err)
+			return []byte{}, err
+		}
+		return clearData, nil
+	}
+	return []byte{}, errors.New("Unsupported Encryption protocol")
+}
+
+func getControllerCertEcdhKey(cipherBlock types.CipherBlock) (*ecdsa.PublicKey, error) {
+	var ecdhPubKey *ecdsa.PublicKey
+	block := cipherBlock.ControllerCert
```

Comment:
The certs will come from the cert device API since they don't change much and are very large.

Question is how the flow changes as you only get an ID aka truncated hash of the cert in the cipherBlock.
Are you planning to return an error if you don't have that ID and retry later after a GET of /certs to get the latest?

We need to work out that flow and structure now.

Reply:
The cert API changes should trigger, reprocessing of existing cipherContexts/cipherBlocks.

Reply:
OK